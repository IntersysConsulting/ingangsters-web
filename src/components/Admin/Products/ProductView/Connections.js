import axios from "axios";
import { API } from "../../../../config";

const getToken = () => localStorage.getItem("token");

export const updateProduct = async (data, form) => {
  const body = {
    _id: data._id,
    name: form.target.name.value,
    description: form.target.description.value,
    price: parseInt(form.target.price.value),
    stock: parseInt(form.target.stock.value),
    image: data.image,
    shippable: data.shippable,
    available: data.available
  };
  const config = {
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + getToken()
    }
  };
  const endpoint = `${API}/products/update`;
  try {
    const res = await axios.put(endpoint, body, config);
    if (res.status) {
      alert("Saved successfully");
      window.location.pathname = "admin/product/" + data._id;
    }
  } catch (err) {
    console.log(err);
  }
};

export const loadProduct = async (id, setData) => {
  const config = {
    headers: {
      "Content-Type": "application/json"
    }
  };
  const endpoint = `${API}/products/single`;
  const body = JSON.stringify({ _id: id });
  try {
    const res = await axios.post(endpoint, body, config);
    const { status, data } = res;
    var loadingNow = status !== 200;
    setData(prev => ({
      ...data.data,
      ...prev,
      showDelete: true,
      isLoading: loadingNow,
      submit: updateProduct,
      _id: id,
      requireImage: false
    }));
  } catch (err) {
    console.log(err);
    window.location.pathname = "productNotFound";
  }
};

export const newProduct = async (data, form) => {
  const config = {
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + getToken()
    }
  };
  const body = {
    name: form.target.name.value,
    description: form.target.description.value,
    price: parseInt(form.target.price.value),
    stock: parseInt(form.target.stock.value),
    image: data.image,
    shippable: true
  };
  const endpoint = `${API}/products/create`;
  try {
    const res = await axios.post(endpoint, body, config);
    if (res.status) {
      alert("New product created successfully");
      window.location.pathname = "admin/product/" + res.data.data;
    }
  } catch (err) {
    alert("Product not saved");
    console.log(err);
  }
};

const uploadImage = async image => {
  const config = {
    headers: {
      "Content-Type": "multipart/form-data",
      Authorization: "Bearer " + getToken(),
      Accept: "application/json"
    }
  };
  const imgBody = new FormData();
  imgBody.append("image", image);
  const endpoint = `${API}/products/images/upload`;
  try {
    const res = await axios.post(endpoint, imgBody, config);
    return res.data.data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

export const imageChange = setData => async evt => {
  var file = evt.target.files[0];
  setData(prev => ({
    ...prev,
    disableSave: true,
    image: "https://media.giphy.com/media/3o7TKtnuHOHHUjR38Y/giphy.gif"
  }));
  let results = await uploadImage(file);
  if (results != null) {
    setData(prev => ({
      ...prev,
      disableSave: false,
      image: results[0]
    }));
  }
};

export const removeProduct = async productID => {
  // Logic removal
  const config = {
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + getToken()
    },
    data: {
      _id: productID
    }
  };
  const endpoint = `${API}/products/delete`;

  try {
    const response = await axios.delete(endpoint, config);
    const { status } = response;
    if (status === 200) {
      alert("Removed successfully");
      window.location.pathname = "admin/dashboard";
    }
  } catch (err) {
    console.log(err);
    alert("An error occurred");
  }
};
