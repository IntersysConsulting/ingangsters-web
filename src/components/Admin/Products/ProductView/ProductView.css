.productForm img {
  display: block;
  width: 100%;
  height: auto;
  margin: auto;
  max-width: 350px;
}

.productForm .btn {
  height: auto;
  display: block;
  margin: auto;
  min-width: 100px;
}

.productForm .bottomButtonBar {
  width: 100%;
  display: flex;
  justify-content: space-around;
}
