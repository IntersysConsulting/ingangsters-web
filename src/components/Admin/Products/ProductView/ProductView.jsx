import React, { useState } from "react";
import Bar from "../../../NavBars/AdminNavBar/AdminNavBar";
import Tabs from "../../AdminTabBar/AdminTabBar";
import Row from "react-bootstrap/Row";
import Container from "react-bootstrap/Container";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import BouncingBall from "../../../UI/Loading/Loading";
import Shadow from "../../../UI/Shadow/Shadow";
import "./ProductView.css";
import { newProduct, loadProduct, imageChange } from "./Connections";

const ProductView = ({ match }) => {
  const { id } = match.params;
  const [data, setData] = useState({
    showDelete: false,
    isLoading: id !== "new",
    submit: newProduct,
    disableSave: false
  });
  let deleteVariant;
  if (data.showDelete) deleteVariant = <Button variant="danger">Delete</Button>;
  else deleteVariant = <span />;

  if (data.isLoading) {
    loadProduct(id, setData);
    return (
      <React.Fragment>
        <Shadow />
        <Bar />
        <Tabs />
        <br />
        <Container>
          <Row className="row adminProductGrid">
            <BouncingBall />
          </Row>
        </Container>
      </React.Fragment>
    );
  } else {
    return (
      <React.Fragment>
        <Shadow />
        <Bar />
        <Tabs />
        <br />
        <Container>
          <h1>{data.name ? data.name : "New Product"}</h1>
          <br />
          <Form
            className="productForm"
            onSubmit={event => {
              event.preventDefault();
              data.submit(data, event);
            }}
            encType="multipart/form-data"
          >
            <Row>
              <Col md={6} sm={12}>
                <Form.Group controlId="name">
                  <Form.Label>Product Name</Form.Label>
                  <Form.Control
                    type="text"
                    placeholder="e.g. Fabuloso"
                    defaultValue={data.name}
                    required
                  />
                </Form.Group>
                <Form.Group controlId="price">
                  <Form.Label>Price</Form.Label>
                  <Form.Control
                    type="number"
                    placeholder="e.g. 5"
                    defaultValue={data.price}
                    required
                    min={1}
                  />
                </Form.Group>
                <Form.Group controlId="stock">
                  <Form.Label>Stock</Form.Label>
                  <Form.Control
                    type="number"
                    placeholder="e.g. 37"
                    defaultValue={data.stock}
                    required
                    min={1}
                  />
                </Form.Group>
                <Form.Group controlId="description">
                  <Form.Label>Description</Form.Label>
                  <Form.Control
                    as="textarea"
                    rows="4"
                    placeholder="This is an awesome product"
                    defaultValue={data.description}
                    required
                  />
                </Form.Group>
              </Col>
              <Col md={6} sm={12} className="mx-auto imageWrapper">
                <img
                  alt="my product"
                  src={
                    data.image
                      ? data.image
                      : "http://mpmco.com/wp-content/uploads/2018/02/placeholder.jpg"
                  }
                  id="productImagePreview"
                />
                <br />
                <Form.Group controlId="image">
                  <Form.Label className="mx-auto btn btn-success">
                    Upload an image
                  </Form.Label>
                  <Form.Control
                    type="file"
                    required={!data.image}
                    className="inputfile"
                    onChange={imageChange(setData)}
                  />
                </Form.Group>
              </Col>
            </Row>
            <br />
            <Row>
              <Col className="bottomButtonBar">
                <Button
                  variant="primary"
                  className="mx-auto"
                  type="submit"
                  disabled={data.disableSave}
                >
                  Save
                </Button>
                {deleteVariant}
              </Col>
              <br />
            </Row>
            <br />
          </Form>
        </Container>
      </React.Fragment>
    );
  }
};

export default ProductView;
