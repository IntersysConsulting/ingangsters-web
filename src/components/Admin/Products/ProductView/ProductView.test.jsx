import React from "react";
import { shallow } from "enzyme";
import ProductView from "./ProductView";
import { createMemoryHistory } from "history";
describe("ADMIN: Products - Product View", () => {
  it('renders "new product" form', () => {
    const history = createMemoryHistory("/admin/product/new");
    const wrapper = shallow(
      <ProductView match={{ params: { id: "new" } }} history={history} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("_default").length).toBe(0);
  });

  it("renders bouncing ball when loading from database", () => {
    const history = createMemoryHistory(
      "/admin/product/5d8e482ed10a23999a8b232d"
    );
    const wrapper = shallow(
      <ProductView
        match={{ params: { id: "5d8e482ed10a23999a8b232d" } }}
        history={history}
      />
    );
    const list = wrapper.find("_default");
    expect(list.length).toBe(1);
    expect(wrapper.find("Form").length).toBe(0);
  });
});
