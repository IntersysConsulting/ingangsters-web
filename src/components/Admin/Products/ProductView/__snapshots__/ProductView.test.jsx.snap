// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ADMIN: Products - Product View renders "new product" form 1`] = `
<Fragment>
  <Connect(NavBarComponent) />
  <Connect(AdminTabBar) />
  <br />
  <Container
    fluid={false}
  >
    <h1>
      New Product
    </h1>
    <br />
    <Form
      className="productForm"
      encType="multipart/form-data"
      inline={false}
      onSubmit={[Function]}
    >
      <ForwardRef
        noGutters={false}
      >
        <Col
          md={6}
          sm={12}
        >
          <FormGroup
            controlId="name"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Product Name
            </FormLabel>
            <FormControl
              placeholder="e.g. Fabuloso"
              required={true}
              type="text"
            />
          </FormGroup>
          <FormGroup
            controlId="price"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Price
            </FormLabel>
            <FormControl
              min={1}
              placeholder="e.g. 5"
              required={true}
              type="number"
            />
          </FormGroup>
          <FormGroup
            controlId="stock"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Stock
            </FormLabel>
            <FormControl
              min={1}
              placeholder="e.g. 37"
              required={true}
              type="number"
            />
          </FormGroup>
          <FormGroup
            controlId="description"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Description
            </FormLabel>
            <FormControl
              as="textarea"
              placeholder="This is an awesome product"
              required={true}
              rows="4"
            />
          </FormGroup>
        </Col>
        <Col
          className="mx-auto imageWrapper"
          md={6}
          sm={12}
        >
          <img
            alt="my product"
            id="productImagePreview"
            src="http://mpmco.com/wp-content/uploads/2018/02/placeholder.jpg"
          />
          <br />
          <FormGroup
            controlId="image"
          >
            <FormLabel
              className="mx-auto btn btn-success"
              column={false}
              srOnly={false}
            >
              Upload an image
            </FormLabel>
            <FormControl
              className="inputfile"
              onChange={[Function]}
              required={true}
              type="file"
            />
          </FormGroup>
        </Col>
      </ForwardRef>
      <br />
      <ForwardRef
        noGutters={false}
      >
        <Col
          className="bottomButtonBar"
        >
          <Button
            active={false}
            className="mx-auto"
            disabled={false}
            type="submit"
            variant="primary"
          >
            Save
          </Button>
          <span />
        </Col>
        <br />
      </ForwardRef>
      <br />
    </Form>
  </Container>
</Fragment>
`;
